### PR TITLE
fix macosx syscalls WIP

### PR DESCRIPTION
**Pull request by jacobrosenthal**
Syscalls are pretty rough on macosx right now. randomly using `r2 /usr/local/bin/go` to test so I can get a wide array of calls, go version go1.10.3 darwin/amd64

```
[0x01057890]> aei ; aeim ; aeip; e asm.emu=1; aa; /s
[x] Analyze all flags starting with sym. and entry0 (aa)
0x010578a9 exit
0x01057912 open
0x0149be0d syscall
0x0172c18a syscall
0x01771c13 syscall
0x0179043e syscall
[0x01057890]> 
```
The numbering is mixing hex and decimal in the sdb.txt at the very least. For instance, 10 only comes after 9 in decimals.. [unlink](https://github.com/radare/radare2/blob/master/libr/syscall/d/darwin-x86-64.sdb.txt#L11-L12) would be 0x200000a not 0x2000010

With these new values generated from https://sigsegv.pl/osx-bsd-syscalls/osx-bsd-syscalls.json with this script
```js
var syscalls = require("./osx-bsd-syscalls.json");

var printCall = function(c,i,o){
	if(c[2] && c[2]!=="enosys" && c[2]!=="nosys"){
		var call = parseInt(c[0]) + 0x2000000;
    	        console.log("%s=0x80,0x%s,%d",c[2], call.toString(16), c[3]);
	}
}

syscalls.forEach(printCall)
```
Now /s gets a bunch more
```
aei ; aeim ; aeip; e asm.emu=1; aa; /s
0x010578a9 exit
0x010578da bsdthread_terminate
0x01057912 open
0x010579b3 getpid
0x010579c6 kill
0x010579f3 setitimer
0x01057a13 madvise
0x01057c33 __pthread_sigmask
0x01057c6d sigaction
0x01057ceb sigreturn
0x01057d12 mmap
0x01057d4f munmap
0x01057d71 sigaltstack
0x01057dd3 select
0x01057e0b bsdthread_create
0x01057ea1 bsdthread_register
0x01057fb2 sysctl
0x01057fea kqueue
0x01058021 kevent
0x01058047 fcntl
```
 but still missing 2/3rds of them?! and Im not sure why. Heres an obvious write call that gets missed, note it also doesnt get the args, also note its mixing eax and rax which is kind of annoying
```
[0x01057890]> s 0x01057973
[0x01057973]> pdf
/ (fcn) sym.runtime.write 33
|   sym.runtime.write (int arg_8h, int arg_10h, int arg_18h, int arg_20h);
|           ; arg int arg_8h @ rsp+0x8
|           ; arg int arg_10h @ rsp+0x10
|           ; arg int arg_18h @ rsp+0x18
|           ; arg int arg_20h @ rsp+0x20
|           ; CALL XREF from sym.runtime.badmorestackg0 (0x102d9dd)
|           ; CALL XREF from sym.runtime.badmorestackgsignal (0x102da1d)
|           ; CALL XREF from sym.runtime.needm (0x103066a)
|           ; CALL XREF from sym.runtime.writeErr (0x1050eed)
|           0x01057960      488b7c2408     mov rdi, qword [arg_8h]     ; [0x8:8]=-1 ; 8 ; rdi=0xffffffffffffffff
|           0x01057965      488b742410     mov rsi, qword [arg_10h]    ; [0x10:8]=-1 ; 16 ; rsi=0xffffffffffffffff
|           0x0105796a      8b542418       mov edx, dword [arg_18h]    ; [0x18:4]=-1 ; 24 ; edx=0xffffffff ; rdx=0xffffffff
|           0x0105796e      b804000002     mov eax, 0x2000004          ; eax=0x2000004 ; rax=0x2000004
|           ;-- hit0_5:
|           0x01057973      0f05           syscall                     ; 33554436 = write (0xffffffffffffffff, 0xffffffffffffffff, 0xffffffff)
|       ,=< 0x01057975      7305           jae 0x105797c               ; unlikely
|       |   0x01057977      b8ffffffff     mov eax, 0xffffffff         ; -1 ; eax=0xffffffff ; rax=0xffffffff
|       `-> 0x0105797c      89442420       mov dword [arg_20h], eax
\           0x01057980      c3             ret                         ; rip=0xffffffffffffffff ; rsp=0xffffffff0683eb46
[0x01057973]> 
````
And heres how many  `/A swi~syscall` finds, a few might be mischaracterized data but Ive gone through a bunch that clearly aren't
```
[0x01057890]> aei ; aeim ; aeip; e asm.emu=1; aa; /A swi~syscall
0x010578a9 2 syscall
0x010578da 2 syscall
0x01057912 2 syscall
0x01057929 2 syscall
0x01057952 2 syscall
0x01057973 2 syscall
0x010579b3 2 syscall
0x010579c6 2 syscall
0x010579f3 2 syscall
0x01057a13 2 syscall
0x01057bf1 2 syscall
0x01057c33 2 syscall
0x01057c6d 2 syscall
0x01057ceb 2 syscall
0x01057d12 2 syscall
0x01057d4f 2 syscall
0x01057d71 2 syscall
0x01057dd3 2 syscall
0x01057e0b 2 syscall
0x01057ea1 2 syscall
0x01057ee8 2 syscall
0x01057f05 2 syscall
0x01057f15 2 syscall
0x01057f29 2 syscall
0x01057f41 2 syscall
0x01057f59 2 syscall
0x01057f7a 2 syscall
0x01057fb2 2 syscall
0x01057fea 2 syscall
0x01058021 2 syscall
0x01058047 2 syscall
0x01077be4 2 syscall
0x01077c4e 2 syscall
0x01077cbf 2 syscall
0x01468445 2 syscall
0x0146ecdd 2 syscall
0x0146edbd 2 syscall
0x01476bd5 2 syscall
0x01476cb5 2 syscall
0x014921cd 2 syscall
0x014971c5 2 syscall
0x014976dd 2 syscall
0x014ccd37 2 syscall
0x014d77a1 2 syscall
0x015265f5 2 syscall
0x0152b5a1 2 syscall
0x0154c97d 2 syscall
0x0154ca92 2 syscall
0x0154ce84 2 syscall
0x0154d5f5 2 syscall
0x0154fbfb 2 syscall
0x0154fe5e 2 syscall
0x0154fe61 2 syscall
0x0154fef8 2 syscall
0x01550069 2 syscall
0x01551994 2 syscall
0x015519a8 2 syscall
0x01551fbc 2 syscall
0x015523b2 2 syscall
0x01554b89 2 syscall
0x01556d57 2 syscall
0x0155c422 2 syscall
0x01562693 2 syscall
0x01568bac 2 syscall
0x01569285 2 syscall
0x0156b3f0 2 syscall
0x01571ed9 2 syscall
0x01578c71 2 syscall
0x01578c7d 2 syscall
0x01578c81 2 syscall
0x01578c85 2 syscall
0x01578c91 2 syscall
0x01578cf9 2 syscall
0x01578d61 2 syscall
0x01580c37 2 syscall
0x01581fc6 2 syscall
0x015821b7 2 syscall
0x01582213 2 syscall
0x01582441 2 syscall
0x0158abc7 2 syscall
0x0158af51 2 syscall
0x0158cfed 2 syscall
0x0158e03b 2 syscall
0x015996cd 2 syscall
0x0159a4f1 2 syscall
0x0159f9e8 2 syscall
0x015a05fc 2 syscall
0x015ae83d 2 syscall
0x015af8ed 3 syscall
0x015afb0f 2 syscall
0x015b13b9 2 syscall
0x015b1a45 2 syscall
0x015bbde0 2 syscall
0x015c0b50 2 syscall
0x015d1816 2 syscall
0x015d216c 2 syscall
0x015d3c5d 2 syscall
0x015dd324 2 syscall
0x015e5d4a 2 syscall
0x015e7dc5 2 syscall
0x015ec568 2 syscall
0x015f252e 2 syscall
0x015fb056 2 syscall
0x015fe2a8 2 syscall
0x0160a6a3 2 syscall
0x016177f2 2 syscall
0x016203db 2 syscall
0x01623fba 2 syscall
0x016274d1 2 syscall
0x01627609 2 syscall
0x01627cd2 2 syscall
0x01636a99 2 syscall
0x0163b183 2 syscall
0x0163b337 2 syscall
0x0163ed96 2 syscall
0x0164f0f4 2 syscall
0x01650a10 2 syscall
0x01650af3 2 syscall
0x01651f3d 2 syscall
0x01651f3f 2 syscall
0x01664ea0 2 syscall
0x01666288 2 syscall
0x01672d91 2 syscall
0x016774dd 2 syscall
0x01679df0 2 syscall
0x0167d024 2 syscall
0x0168645c 2 syscall
0x016881db 2 syscall
0x0168bd56 2 syscall
0x0168d579 2 syscall
0x016913dc 2 syscall
0x01692548 2 syscall
0x0169cd8a 2 syscall
0x016a0ddc 2 syscall
0x016a39b2 2 syscall
0x016a3e01 2 syscall
0x016aecfd 2 syscall
0x016b4fd1 2 syscall
0x016b5084 2 syscall
0x016b521b 2 syscall
0x016b533a 2 syscall
0x016bde60 2 syscall
0x016cff40 2 syscall
0x016d2482 2 syscall
0x016e8e3a 2 syscall
0x016e9cfc 2 syscall
0x0172a0b6 2 syscall
0x017955d9 2 syscall
0x017c41d8 2 syscall
```

Help wanted

### CODE REVIEW

**Comment by jacobrosenthal** [comment]
Added work from @ stek29  from https://gist.github.com/stek29/08ca323506bcae78673de2e564e105d1

**Comment by radare** [comment]
cool!

**Comment by jacobrosenthal** [comment]
Would still love comment from maybe @sivaramaaa on why /s is still missing a lot of obvious calls like the write I posted above. Is it in the /s command or something about the way were defining the mapping

File: libr/syscall/d/darwin-x86-32.sdb.txt [hunk 1]
```diff
@@ -1,107 +1,468 @@
...
-mk_timer_create_trap=0x81,91,1,
-mk_timer_destroy_trap=0x81,92,1,
-mk_timer_arm_trap=0x81,93,1,
-mk_timer_cancel_trap=0x81,94,1,
-mk_timebase_info_trap=0x81,95,1,
-iokit_user_client_trap=0x81,100,1,
+exit=0x80,0x1,1,
+fork=0x80,0x2,0,
+read=0x80,0x3,3,
+write=0x80,0x4,3,
```
Comment (sivaramaaa) [inline libr/syscall/d/darwin-x86-32.sdb.txt]: This new addition is missing arguments type , so say for write syscall it should be 
```
 write=0x80,4,3,iZi 
```

**Comment by sivaramaaa** [comment]
Apart from my one comment ,this thing lgtm , good work !

Regarding the reason from `/s` not picking up syscalls , i have to investigate !

**Comment by sivaramaaa** [comment]
Also , i think there is one regression that we should take care of 

```
[XX] db/anal/others ahi S  43
$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -i /var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/tmp-29194wz2viZ8YUnh2.tmp malloc://16
e asm.arch=x86
e asm.bits=32
wx b801000000
ahi S
pi 1
-mov eax, exit()
+mov eax, unknown()
```

**Comment by jacobrosenthal** [comment]
Some tiny bit more info. Its not r_syscall_get thats failing. Some printf logging from inside do_syscall_search finds that:

A Good instruction, offset comes back properly from emulateSyscallPrelude and then matches syscall
```
0x010578a9 aop.val=18446744073709551615 curpc=17135771 off=33554433 nbytes=16 nb_opcodes=0 previnstr[curpos]=2
...
[0x01057890]> s 0x010578a9
[0x010578a9]> pdf
/ (fcn) sym.runtime.exit 23
|   sym.runtime.exit (int arg_8h);
|           ; arg int arg_8h @ rsp+0x8
|           ; XREFS: CALL 0x01003640  CALL 0x01026eab  CALL 0x01026f62  CALL 0x0102b614  CALL 0x0102b620  CALL 0x0102b670  
|           ; XREFS: CALL 0x0102b91e  CALL 0x0102cc35  CALL 0x01030676  CALL 0x010394f6  CALL 0x0103f183  CALL 0x0104025d  
|           0x010578a0      8b7c2408       mov edi, dword [arg_8h]     ; [0x8:4]=-1 ; 8 ; edi=0xffffffff ; rdi=0xffffffff
|           0x010578a4      b801000002     mov eax, 0x2000001          ; eax=0x2000001 ; rax=0x2000001
|           ;-- hit0_0:
|           0x010578a9      0f05           syscall                     ; 33554433 = exit (0xffffffff)
|           0x010578ab      c70425f10000.  mov dword [0xf1], 0xf1      ; [0xf1:4]=-1 ; 241
\           0x010578b6      c3             ret                         ; rip=0xffffffffffffffff ; rsp=0xffffffff0683eb46
```




A bad instruction, offset comes back from emulateSyscallPrelude as -2 and then doesnt match syscall
```
0x01057929 aop.val=18446744073709551615 curpc=17135897 off=-2 nbytes=18 nb_opcodes=0 previnstr[curpos]=2 seems likley. curpc = at - (nbytes - previnstr[curpos]);
...
[0x010578a9]> s 0x01057929
[0x01057929]> pdf
/ (fcn) sym.runtime.closefd 23
|   sym.runtime.closefd (int arg_8h, int arg_10h);
|           ; arg int arg_8h @ rsp+0x8
|           ; arg int arg_10h @ rsp+0x10
|           ; CALL XREF from sym.runtime.getRandomData (0x1028e30)
|           0x01057920      8b7c2408       mov edi, dword [arg_8h]     ; [0x8:4]=-1 ; 8 ; edi=0xffffffff ; rdi=0xffffffff
|           0x01057924      b806000002     mov eax, 0x2000006          ; eax=0x2000006 ; rax=0x2000006
|           ;-- hit0_3:
|           0x01057929      0f05           syscall                     ; 33554438 = close (0xffffffff)
|       ,=< 0x0105792b      7305           jae 0x1057932               ; unlikely
|       |   0x0105792d      b8ffffffff     mov eax, 0xffffffff         ; -1 ; eax=0xffffffff ; rax=0xffffffff
|       `-> 0x01057932      89442410       mov dword [arg_10h], eax
\           0x01057936      c3             ret                         ; rip=0xffffffffffffffff ; rsp=0xffffffff0683eb46
[0x01057929]> 
```

**Comment by radare** [comment]
looks red for 1 test:

```
[31m[XX][39m db/anal/others [33mahi S[39m  43


$ r2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Q -i /var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/tmp-29194wz2viZ8YUnh2.tmp malloc://16



e asm.arch=x86

e asm.bits=32

wx b801000000

ahi S

pi 1





[35m-mov eax, exit()[39m

[35m[39m[32m+mov eax, unknown()[39m
```

**Comment by radare** [comment]
fails in this test.

**Comment by radare** [comment]
i think the test fails because its missing an asm.os=linux, and the test is running in a mac

**Comment by radare** [comment]
@Maijin ?

**Comment by Maijin** [comment]
Yep must be this

**Comment by radare** [comment]
Can you fix the test? Is the pr ready to merge (aka finished), lgtm.

**Comment by sivaramaaa** [comment]
@radare yup once the test is fixed , we can merge this , rest can be taken care in another PR !

**Comment by Maijin** [comment]
fixed - https://github.com/radare/radare2-regressions/commit/d3f49c978c75240c9025be3b3de4476d6c56518a

**Comment by radare** [comment]
thanks!

**Pull request merged**